File: src/main/java/dev/ddlproxy/service/JDownloaderController.java
package dev.ddlproxy.service;

import tools.jackson.databind.JsonNode;
import tools.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Map;

public class JDownloaderController {
    private final String JDOWNLOADER_API_URL;

    private final RestTemplate restTemplate = new RestTemplate();
    private static final Logger logger = LoggerFactory.getLogger(JDownloaderController.class);

    public JDownloaderController(String baseURL) {
        this.JDOWNLOADER_API_URL = baseURL;

        restTemplate.getInterceptors().add((request, body, execution) -> {
            logger.trace("Request URI: {}", request.getURI());
            logger.trace("Request Method: {}", request.getMethod());
            logger.trace("Request Headers: {}", request.getHeaders());
            logger.trace("Request Body: {}", new String(body, StandardCharsets.UTF_8));
            return execution.execute(request, body);
        });
    }

    public boolean isLinkOnline(ArrayList<String> link) {
        addLink(link.toString());

        String result;
        int maxRetries = 100;
        int attempt = 0;

        do {
            result = queryLinks();
            attempt++;

            switch (result) {
                case "ONLINE":
                    clearList();
                    return true;
                case "OFFLINE":
                    clearList();
                    return false;
                case "MIXED":
                    logger.trace("Some links are not fully online, continuing to wait...");
                    break;
                case "UNKNOWN":
                    logger.trace("Query returned UNKNOWN state, retrying...");
                    break;
                case "ERROR":
                    logger.error("Error encountered while querying links.");
                    break;
                default:
                    logger.warn("Unexpected link state: {}", result);
                    break;
            }

            logger.trace("Query result is UNKNOWN, retrying...");

            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }

        } while (attempt < maxRetries);
        clearList();
        logger.warn("maxRetries reached on link \"{}\"", link);
        return false;
    }

    private void addLink(String link) {
        logger.debug("Adding link \"{}\"", link);
        String url = JDOWNLOADER_API_URL + "/linkgrabberv2/addLinks";

        Map<String, Object> payload = Map.of(
                "links", link,
                "autostart", false,
                "url", true,
                "enabled", true
        );

        restTemplate.postForEntity(
                url,
                new HttpEntity<>(payload, new HttpHeaders() {{
                    setContentType(MediaType.APPLICATION_JSON);
                }}),
                Void.class
        );
    }

    private String queryLinks() {
        String url = JDOWNLOADER_API_URL + "/linkgrabberv2/queryLinks";

        Map<String, Object> payload = Map.of(
                "availability", true,
                "enabled", true
        );

        try {
            ResponseEntity<String> responseEntity = restTemplate.postForEntity(
                    url,
                    new HttpEntity<>(payload, new HttpHeaders() {{
                        setContentType(MediaType.APPLICATION_JSON);
                    }}),
                    String.class
            );

            String jsonResponse = responseEntity.getBody();
            logger.trace("payload: {}", payload);
            logger.trace("response: {}", jsonResponse);

            JsonNode rootNode = new ObjectMapper().readTree(jsonResponse);
            JsonNode dataArray = rootNode.path("data");

            if (!dataArray.isArray() || dataArray.isEmpty()) {
                logger.debug("No data returned from linkgrabber query");
                return "UNKNOWN";
            }

            boolean anyOffline = false;
            boolean allOnline = true;

            for (JsonNode data : dataArray) {
                String availability = data.path("availability").asString();
                if (!availability.equalsIgnoreCase("ONLINE")) {
                    allOnline = false;
                    if (availability.equalsIgnoreCase("OFFLINE")) {
                        anyOffline = true;
                    }
                }
            }

            if (allOnline) return "ONLINE";
            if (anyOffline) return "OFFLINE";
            return "MIXED";

        } catch (Exception e) {
            logger.error("Error during queryLinks", e);
            return "ERROR";
        }
    }

    private void clearList() {
        String url = JDOWNLOADER_API_URL + "/linkgrabberv2/clearList";

        ResponseEntity<String> response = restTemplate.getForEntity(url, String.class);

        if (!response.getStatusCode().is2xxSuccessful()) {
            logger.error("Failed to clear linkgrabber list: {}", response.getBody());
            throw new RuntimeException("Clear list failed: " + response.getStatusCode());
        }
        logger.debug("Linkgrabber list cleared successfully");
    }

    public void download(ArrayList<String> link, String destinationFolder) {
        clearList();

        Map<String, Object> payload = Map.of(
                "links", String.join("\n", link),
                "autostart", true,
                "destinationFolder", destinationFolder,
                "enabled", true,
                "autoExtract", true,
                "overwritePackagizerRules", true
        );

        try {
            ResponseEntity<String> jdownloaderResponse = restTemplate.postForEntity(
                    JDOWNLOADER_API_URL + "/linkgrabberv2/addLinks",
                    new HttpEntity<>(payload, new HttpHeaders() {{
                        setContentType(MediaType.APPLICATION_JSON);
                    }}),
                    String.class
            );

            logger.trace("JDownloader response: {}", jdownloaderResponse.getBody());
            logger.info("Started download: {}", link);
        } catch (Exception e) {
            logger.error("Failed to start download: {}", link, e);
        }
    }
}

File: src/main/kotlin/dev/ddlproxy/service/JDownloaderController.kt
package dev.ddlproxy.service

import dev.ddlproxy.model.LinkGroup
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import tools.jackson.databind.ObjectMapper
import org.springframework.http.HttpStatus

sealed class LinkState {
    object Online : LinkState()
    object Offline : LinkState()
    object Mixed : LinkState()
    object Unknown : LinkState()
    object Error : LinkState()
}

@Service
class JDownloaderController(
    private val restTemplate: RestTemplate,
    private val objectMapper: ObjectMapper,
    @Value($$"${jdownloader.api.url}") jdownloaderApiUrlRaw: String,
    @param:Value($$"${download.folder}") private val downloadFolder: String
) {

    private val logger = LoggerFactory.getLogger(JDownloaderController::class.java)

    private val jdownloaderApiUrl = jdownloaderApiUrlRaw.trim().trimEnd('/')

    fun isLinkOnline(links: LinkGroup): Boolean {
        addLinks(links.links)

        var result: LinkState
        var attempt = 0
        val maxRetries = 100

        do {
            result = queryLinks()
            attempt++

            when (result) {
                is LinkState.Online -> {
                    clearList()
                    return true
                }

                is LinkState.Offline -> {
                    clearList()
                    return false
                }

                is LinkState.Mixed -> {
                    logger.trace("Some links are not fully online, continuing to wait...")
                }

                is LinkState.Unknown -> {
                    logger.trace("Query returned UNKNOWN state, retrying...")
                }

                is LinkState.Error -> {
                    logger.error("Error encountered while querying links.")
                    break
                }
            }

            logger.trace("Query result is UNKNOWN, retrying...")

            try {
                Thread.sleep(1000)
            } catch (e: InterruptedException) {
                Thread.currentThread().interrupt()
            }

        } while (attempt < maxRetries)

        clearList()
        logger.warn("Max retries reached on links: {}", links)
        return false
    }

    private fun addLinks(links: List<String>) {
        logger.debug("Adding links: {}", links)

        val url = "$jdownloaderApiUrl/linkgrabberv2/addLinks"

        val payload = mapOf(
            "links" to links.joinToString("\n"),
            "autostart" to false,
            "url" to true,
            "enabled" to true
        )

        restTemplate.postForEntity(
            url,
            HttpEntity(payload, HttpHeaders().apply { contentType = MediaType.APPLICATION_JSON }),
            Void::class.java
        )
    }

    private fun queryLinks(): LinkState {
        val url = "$jdownloaderApiUrl/linkgrabberv2/queryLinks"

        val payload = mapOf(
            "availability" to true,
            "enabled" to true
        )

        return try {
            val responseEntity: ResponseEntity<String> = restTemplate.postForEntity(
                url,
                HttpEntity(payload, HttpHeaders().apply { contentType = MediaType.APPLICATION_JSON }),
                String::class.java
            )

            val jsonResponse = responseEntity.body ?: return LinkState.Unknown
            logger.trace("Response: {}", jsonResponse)

            val rootNode = objectMapper.readTree(jsonResponse)
            val dataArray = rootNode.path("data")

            if (!dataArray.isArray || dataArray.isEmpty) {
                logger.debug("No data returned from linkgrabber query")
                return LinkState.Unknown
            }

            var anyOnline = false
            var anyOffline = false

            for (data in dataArray) {
                val availability = data.path("availability").asString()

                when {
                    availability.equals("ONLINE", ignoreCase = true) -> anyOnline = true
                    availability.equals("OFFLINE", ignoreCase = true) -> anyOffline = true
                }
            }

            return when {
                anyOnline && !anyOffline -> LinkState.Online
                !anyOnline && anyOffline -> LinkState.Offline
                anyOnline && anyOffline -> LinkState.Mixed
                else -> LinkState.Unknown
            }

        } catch (e: Exception) {
            logger.error("Error during queryLinks", e)
            LinkState.Error
        }
    }

    private fun clearList() {
        val url = "$jdownloaderApiUrl/linkgrabberv2/clearList"

        val response: ResponseEntity<String> = restTemplate.getForEntity(url, String::class.java)

        if (response.statusCode != HttpStatus.OK) {
            logger.error("Failed to clear linkgrabber list: {}", response.body)
            throw RuntimeException("Clear list failed: ${response.statusCode}")
        }

        logger.debug("Linkgrabber list cleared successfully")
    }

    /**
     * Starts downloading the given [LinkGroup] if online.
     *
     * @param linkGroup The link group to download.
     * @param skipOnlineCheck Do not check if link is online before attempting download
     *
     * Use [download] with a [List] of [LinkGroup]s if you have multiple and want
     * to automatically pick the first online link group.
     */
    fun download(linkGroup: LinkGroup, skipOnlineCheck: Boolean = false) {
        if (!skipOnlineCheck && !isLinkOnline(linkGroup)) {
            logger.warn("No valid links found for {}", linkGroup)
            return
        }

        clearList()

        val payload = mapOf(
            "links" to linkGroup.links.joinToString("\n"),
            "autostart" to true,
            "destinationFolder" to downloadFolder,
            "enabled" to true,
            "autoExtract" to true,
            "overwritePackagizerRules" to true
        )

        try {
            val response: ResponseEntity<String> = restTemplate.postForEntity(
                "$jdownloaderApiUrl/linkgrabberv2/addLinks",
                HttpEntity(payload, HttpHeaders().apply { contentType = MediaType.APPLICATION_JSON }),
                String::class.java
            )

            logger.trace("JDownloader response: {}", response.body)
            logger.info("Started download: {}", linkGroup)
        } catch (e: Exception) {
            logger.error("Failed to start download: {}", linkGroup, e)
        }
    }

    /**
     * Starts downloading the first online [LinkGroup] in the given list.
     *
     * @param linkGroups The list of link groups to try. This should be sorted in the order you want them checked.
     *
     * Use [download] with a single [LinkGroup] if you only have one to download.
     */
    fun download(linkGroups: List<LinkGroup>) {
        val link = linkGroups.firstOrNull { isLinkOnline(it) } ?: run {
            logger.warn("No valid links found for {}", linkGroups)
            return
        }

        download(link)
    }
}
